### ISSUE TITLE
Nested `attrs` result in the wrong prop value when one prop depends on another

### ISSUE THREAD

**Issue by exogen** [styled-components/styled-components#3178]
## Environment

`npx envinfo --system --binaries --npmPackages styled-components,babel-plugin-styled-components --markdown --clipboard`

```
## System:
 - OS: macOS 10.15.5
 - CPU: (8) x64 Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz
 - Memory: 611.90 MB / 16.00 GB
 - Shell: 3.2.57 - /bin/bash
## Binaries:
 - Node: 10.13.0 - ~/.nvm/versions/node/v10.13.0/bin/node
 - Yarn: 1.22.4 - /usr/local/bin/yarn
 - npm: 6.10.2 - ~/.nvm/versions/node/v10.13.0/bin/npm
## npmPackages:
 - styled-components: ^5.1.1 => 5.1.1
```

## Reproduction

https://codesandbox.io/s/nostalgic-dhawan-y2jiu?file=/src/App.js

## Steps to reproduce

Use `attrs` in a nested way. Let's start with this component:

```js
const Foo = styled.button.attrs(props => ({
  size: props.size || 10
})`
  font-size: ${props => props.size}px;
`;
```

Then `<Foo>` will render a button with font-size 10. `<Foo size={20}>` will render a button with font-size 20. So far so good!

Now let's create another styled-component from that:

```js
const Bar = styled(Foo).attrs({
  size: 20
})`
  color: red;
`;
```

Now rendering `<Bar>` will render a button with font-size 20. Still all good!

Now let's update Foo to default one of its *other* props to depend on the `size` prop:

```js
const Foo = styled.button.attrs(props => ({
  size: props.size || 10,
  children: props.children || `size: ${props.size}`
})`
  font-size: ${props => props.size}px;
`;
```

Now something strange happens. `<Bar>` doesn't pass `children`, but does pass `size: 20` – so we expect Foo's `children` default to run and see `size: 20`. But when we render `<Bar>` we see:

```
size: undefined
```

It seems as if somehow the `children` default from Foo was applied *before* Bar's `attrs` were taken into consideration.

Rendering `<Bar size={20}>` does give `children` the correct value. The result is that rendering `<Bar>` which is defined as `styled(Foo).attrs({ size: 20 })` differs from rendering `<Foo size={20}>`.

If we add a `console.log` in Foo's `attrs` function, we see it's getting `size: 20` correctly, but is also receiving `children: undefined`. Somehow, its own default has already been applied even though we should only have applied Bar's `attrs` at this point.

## Expected Behavior

```jsx
const Bar = styled(Foo).attrs({
  size: 20
})``;

<Bar />
```

Should result in the same output as:

```jsx
<Foo size={20} />
```

## Actual Behavior

The `Expected Behavior` output is different for the two cases.

There seems to be some odd merging behavior between `attrs` which breaks the model of `styled(Whatever)` being a higher-order component. If it were a true higher order component then this would work, since `Foo` would get `children` as undefined and thus apply its default.

**Comment by kitten** [comment]
This happens because `attrs` are applied in reverse-order, meaning inside-our, rather than outside-in, so by the time `children.size` is read it hasn’t been applied yet by the outside wrapping `StyledComponent`.

One thing I always add is that styled-components are supposed to not be used as layers of indirection, so keeping them flat is obviously always the preferred method. This applies to all their logic; as soon as we have interacting styles, rules, and also `attrs`, we’re in a position where the general idea of `StyledComponent`s guiding best practices and reducing the mental overhead of writing styles doesn’t apply anymore. But addressing this first, that’s obviously besides the point.

The reason why we don’t apply `attrs` in “natural order”, if I remember this correctly, is because we don’t treat folded components the same as HOCs, both in code and conceptually.

Rather we see `StyledComponent`s that wrap another as “overrides” that extend their corresponding inner components. This is derived from the older `.extend()` API (or whatever it was called 😅)

Hence the mentality is that not only the styles on the outside are applied *after* and on top of the inner component’s existing styles, but also all other compositions, like the `attrs` function.

This is hence intentional behaviour, which we also can’t arbitrarily flip now even if we started considering it unintentional (the last time we did this to an API bad things happened in our issues 😂)

So in summary: If props need to be defaulted and composed, then `attrs`, which by its name is supposed to inform only an underlying structural component’s (depth 1) props or an elements attributes, isn’t a replacement for props composition in other presentational components.

@probablyup: Does that sound about right to you? Anything I’ve missed?

**Comment by kitten** [comment]
Closing due to inactivity, although I'd be happy to revisit this for a major bump.

**Comment by vzaidman** [comment]
for when you want to revisit it:

i was really confusing by this and it took me a hours to realize what's going on. 

what i had was a component enforcing using "alt" for a11y reasons or an explicit noAlt:
```
// BAD CODE BELOW
export const ImageWithAlt = styled.img.attrs(({noAlt, alt, src, displayName}) => {
  if (!alt && !noAlt) {
    logger.errorDialog('no alt was provided to ImageWithAlt, use noAlt attr incase of decorative images.', {component: displayName});
  }

  return {
    alt: noAlt ? '' : alt,
    src: src || minimalGifSrc,
  };
})``;
```
and it was used like this:
```
const Image = styled(ImageWithAlt).attrs({
  src: noResultImageSrc,
  alt: 'No result...',
})`
  height: 150px;
  width: 210px;
`;
```

I think an `.override` syntax would of clarify that there's no HOC behavior here:
```
const Image = styled.override(ImageWithAlt).attrs({
  alt: 'No result...',
})`
...
`;
```

but such a major change to the API deferentially sounds too much.

what about adding a line in the code that informs the user about the order in which the attrs are applied with a link to docs?
![image](https://user-images.githubusercontent.com/5188459/91716976-4a761e00-eb99-11ea-89dc-315af3463187.png)

**Comment by vzaidman** [comment]
@kitten I opened a PR for adding a link to docs explaining that .attrs are resolved bottom-to-top to mitigate this issue:
https://github.com/styled-components/styled-components/pull/3299
